src/Interpreters/convertFieldToType.cpp

```diff
@@ -243,9 +243,9 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
             return dynamic_cast<const IDataTypeEnum &>(type).castToValue(src);
         }
 
-        if ((which_type.isDate() || which_type.isDateTime()) && src.getType() == Field::Types::UInt64)
+        if ((which_type.isDate() || which_type.isDateTime()) && (src.getType() == Field::Types::UInt64 || src.getType() == Field::Types::Decimal64))
         {
-            /// We don't need any conversion UInt64 is under type of Date and DateTime
+            /// We don't need any conversion as both Decimal64 and UInt64 are under types of Date and DateTime
```

Comment:
Can you explain this change? It's not trivial to me why Decimal64 value should not be converted to Date/DateTime

Reply:
I would say Decimal64 is not convertable to Date or DateTime

Reply:
query:
```
SELECT ts FROM file('data.parquet', Parquet, 'ts DateTime')
```

error:
```
Code: 53. DB::Exception: Type mismatch in IN or VALUES section. Expected: DateTime. Got: Decimal64:  (When parsing Parquet statistics for column ts, physical type 2, column descriptor = {
  name: ts,
  path: ts,
  physical_type: INT64,
  converted_type: NONE,
  logical_type: Timestamp(isAdjustedToUTC=false, timeUnit=microseconds, is_from_converted_type=false, force_set_converted_type=false),
  max_definition_level: 1,
  max_repetition_level: 0,
}. Please report an issue and use input_format_parquet_filter_push_down = false to work around.): (in file/uri /home/ubuntu/data.parquet): While executing ParquetBlockInputFormat: While executing File. (TYPE_MISMATCH)
```

introduced after: https://github.com/ClickHouse/ClickHouse/pull/68131

and this change fixed it, so it looks like `Decimal64` is convertable to at least `DateTime `

Reply:
> and this change fixed it, so it looks like Decimal64 is convertable to at least DateTime 

It fixed the exception but might lead to the wrong result. This error happens during reading statistics from Parquet metadata file to apply min/max statistics for filter pushdown oprimization. And when real value type is `DateTime64(6)` (as in your example) and we request it as `DateTime` we may lose precision during conversion of the min/max values.
So, it would be better to:
- Check this type of conversion before calling `convertFieldToType` in `ParquetBlockInputFormat.cpp::getHyperrectangleForRowGroup` and disable index (use default Field{} for min/max)
- Consider this conversion as valid in this particular case in Parquet and do conversion explicitely in `ParquetBlockInputFormat.cpp::getHyperrectangleForRowGroup` instead of calling `convertFieldToType`.

I am actually not 100% sure if such conversion can lead to unexpected results. Let's ask @al13n321 for an advice here

Reply:
Similar exception will be during conversion of `DateTime64(6)` to `Int64`:
```sql
:) SELECT last_purchase_date
FROM file('segments_customer_attributes_v2_us_neweracap_20240915T100000_000000000001.parquet', Parquet, 'last_purchase_date Int64')
LIMIT 1

Query id: f835d84c-a284-4df4-965a-507c5fdf76d2


Elapsed: 0.007 sec.

Received exception:
Code: 53. DB::Exception: Type mismatch in IN or VALUES section. Expected: Int64. Got: Decimal64:  (When parsing Parquet statistics for column last_purchase_date, physical type 2, column descriptor = {
  name: last_purchase_date,
  path: last_purchase_date,
  physical_type: INT64,
  converted_type: NONE,
  logical_type: Timestamp(isAdjustedToUTC=false, timeUnit=microseconds, is_from_converted_type=false, force_set_converted_type=false),
  max_definition_level: 1,
  max_repetition_level: 0,
}. Please report an issue and use input_format_parquet_filter_push_down = false to work around.): (in file/uri /home/avogar/tmp/data/segments_customer_attributes_v2_us_neweracap_20240915T100000_000000000001.parquet): While executing ParquetBlockInputFormat: While executing File. (TYPE_MISMATCH)
```

Probably, we should use `try/catch` around `convertFieldToType` and disable index in this case instead of throwing an exception. Or implement function `tryConvertFieldToType` that will not throw an exception